src/drivers/pwm_input/pwm_input.cpp

```diff
@@ -314,160 +214,43 @@ void PWMIN::_timer_init(void)
+	/* enable interrupts */
+	up_enable_irq(PWMIN_TIMER_VECTOR);
 }
 
-// XXX refactor this out of this driver
 void
-PWMIN::hard_reset()
-{
-	_turn_off();
-	hrt_call_after(&_hard_reset_call, 9000, reinterpret_cast<hrt_callout>(&PWMIN::_turn_on), this);
-}
-
-/*
- * hook for open of the driver. We start the timer at this point, then
- * leave it running
- */
-int
-PWMIN::open(struct file *filp)
-{
-	if (g_dev == nullptr) {
-		return -EIO;
-	}
-
-	int ret = CDev::open(filp);
-
-	if (ret == OK && !_timer_started) {
-		g_dev->_timer_init();
-	}
-
-	return ret;
-}
-
-
-/*
- * handle ioctl requests
- */
-int
-PWMIN::ioctl(struct file *filp, int cmd, unsigned long arg)
-{
-	switch (cmd) {
-	case SENSORIOCRESET:
-		/* user has asked for the timer to be reset. This may
-		 * be needed if the pin was used for a different
-		 * purpose (such as PWM output) */
-		_timer_init();
-		/* also reset the sensor */
-		hard_reset();
-		return OK;
-
-	default:
-		/* give it to the superclass */
-		return CDev::ioctl(filp, cmd, arg);
-	}
-}
-
-
-/*
- * read some samples from the device
- */
-ssize_t
-PWMIN::read(struct file *filp, char *buffer, size_t buflen)
-{
-	_last_read_time = hrt_absolute_time();
-
-	unsigned count = buflen / sizeof(struct pwm_input_s);
-	struct pwm_input_s *buf = reinterpret_cast<struct pwm_input_s *>(buffer);
-	int ret = 0;
-
-	/* buffer must be large enough */
-	if (count < 1) {
-		return -ENOSPC;
-	}
-
-	while (count--) {
-		if (_reports->get(buf)) {
-			ret += sizeof(struct pwm_input_s);
-			buf++;
-		}
-	}
-
-	/* if there was no data, warn the caller */
-	return ret ? ret : -EAGAIN;
-}
-
-/*
- * publish some data from the ISR in the ring buffer
- */
-void PWMIN::publish(uint16_t status, uint32_t period, uint32_t pulse_width)
+PWMIN::publish(uint16_t status, uint32_t period, uint32_t pulse_width)
 {
 	/* if we missed an edge, we have to give up */
 	if (status & SR_OVF_PWMIN) {
 		_error_count++;
 		return;
 	}
 
-	_last_poll_time = hrt_absolute_time();
+	_pwm.timestamp = hrt_absolute_time();
+	_pwm.error_count = _error_count;
+	_pwm.period = period;
+	_pwm.pulse_width = pulse_width;
 
-	struct pwm_input_s pwmin_report;
-	pwmin_report.timestamp = _last_poll_time;
-	pwmin_report.error_count = _error_count;
-	pwmin_report.period = period;
-	pwmin_report.pulse_width = pulse_width;
+	_pwm_input_pub.publish(_pwm);
 
-	_reports->force(&pwmin_report);
+	// update statistics
+	_last_period = period;
+	_last_width = pulse_width;
+	_pulses_captured++;
 }
 
-/*
- * print information on the last captured
- */
-void PWMIN::print_info(void)
+void
+PWMIN::print_info(void)
 {
-	if (!_timer_started) {
-		printf("timer not started - try the 'test' command\n");
-
-	} else {
-		printf("count=%u period=%u width=%u\n",
-		       (unsigned)_pulses_captured,
-		       (unsigned)_last_period,
-		       (unsigned)_last_width);
-	}
+	PX4_INFO("count=%u period=%u width=%u\n",
+		 (unsigned)_pulses_captured,
+		 (unsigned)_last_period,
+		 (unsigned)_last_width);
```

Comment:
Looks good so far @dakejahl.  Using `static_cast<unsigned int>(var_name)` will allow compile time checking where the current c-style cast will not.